polaris-service/src/main/java/io/polaris/service/catalog/FileIOWrapperFactory.java

```diff
@@ -0,0 +1,32 @@
+/*
+ * Copyright (c) 2024 Snowflake Computing Inc.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package io.polaris.service.catalog;
+
+import com.fasterxml.jackson.annotation.JsonTypeInfo;
+import io.dropwizard.jackson.Discoverable;
+import io.polaris.core.monitor.PolarisMetricRegistry;
+import org.apache.iceberg.io.FileIO;
+
+/** Interface for providing a wrapper to intercept file IO, such as reading/writing S3. */
+@JsonTypeInfo(
+    use = JsonTypeInfo.Id.NAME,
+    include = JsonTypeInfo.As.PROPERTY,
+    property = "wrapperType")
+public interface FileIOWrapperFactory extends Discoverable {
+  void setMetricRegistry(PolarisMetricRegistry metricRegistry);
```

Comment:
This makes sense on some kind of a `MetricRecordingFileIOFactory` rather than defining it on the interface itself